internal/core/src/fns/fns.hpp

```diff
@@ -0,0 +1,297 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef INTERNAL_CORE_SRC_FNS_FNS_HPP_
+#define INTERNAL_CORE_SRC_FNS_FNS_HPP_
+
+#include <memory>
+#include <iostream>
+#include <unordered_set>
+#include <algorithm>
+#include <queue>
+#include <utility>
+#include <vector>
+
+#include "kmeans.h"
+#include "kgraph.h"
+#include "basicDistance.h"
+#include "visitedList.hpp"
+
+#include "common/Types.h"
+using namespace milvus;
+
+using PriorQ = std::priority_queue<std::pair<distance_t, idx_t>,
+                                   std::vector<std::pair<distance_t, idx_t>>>;
+using NNDIndexParam = IndexParams;
+using graph_t = std::vector<std::vector<idx_t>>;
+
+struct BuildParam {
+    size_t kms_cluster_num = 1024;
+    size_t filter_pos = 4;
+    size_t filter_density = 0;
+    size_t filter_ctrl_size = 64;
+    size_t filter_max_edges = 2048;
+};
+
+template <typename T>
+class FNS {
+ private:
+    const T* Data_;
+
+    float* KmeansCentroids_ = nullptr;
+    int* KmeansLabels_ = nullptr;
+
+    graph_t KmeansLabelGraph_;
+    std::vector<idx_t> getDensity_;
+    graph_t kNNGraph_;
+    const size_t dataSz_;
+    const size_t dataDim_;
+    BuildParam BP_;
+    NNDIndexParam nnd_para_;
+
+    std::vector<idx_t> fullLowDenPoints_;
+    size_t searchK_ = 0;
+    VisitedListPool* visited_list_pool_;
```

Comment:
Naming conventions need to be standardized, including camel case, snake case, and capitalization. It's quite chaotic at the moment.